include/seqan3/alignment/pairwise/policy/all.hpp

```diff
@@ -0,0 +1,132 @@
+ *
+ * ## Gap policies
+ *
+ * Gap policies are used to initialise and to compute the cells within the alignment matrix.
+ * The gap policies are further divided into a policy initialising the matrix and computing the cells.
+ * The following table shows the functions that need to be implemented by a gap policy.
+ *
+ * Function name     | Arguments                              | Return value                         |
+ * ----------------- | -------------------------------------- | ------------------------------------ |
+ * `compute_cell`    | `cell &&`, `cache &`, `score const &`  | `void`                               |
+ * `make_cache`      | `gap_scheme const &`                   | alignment_algorithm_cache            |
+ *
+ * - compute_cell:
+ *
+ *    This function implements the kernel that computes the score and, if enabled, the traceback direction.
+ *    It gets as an input the dereferenced value of the scoring matrix (see above), the current alignment algorithm
+ *    cache and the score of comparing two letter of the used alphabet using the passed scoring scheme.
+ *
+ * - make_cache:
+ *
+ *     Initialises and returns the cache used in the alignment algorithm. It must return a type that allows structural
+ *     bindings, with the first argument being of the same type as the allocator value type used in the matrix policy,
+ *     the second argument being the costs for gap opening + gap and the third argument being the costs for a gap.
+ *
+ * ### Existing gap policies:
+ *
+ *  - seqan3::detail::affine_gap_policy
+ *
+ * The following table displays requirements for the corresponding gap intialisation policy:
+ *
+ * Function name      | Arguments                        | Return value                         |
+ * ------------------ | -------------------------------- | ------------------------------------ |
+ * `init_origin_cell` | `cell &&`, `cache &`             | `void`                               |
+ * `init_column_cell` | `cell &&`, `cache &`             | `void`                               |
+ * `init_row_cell`    | `cell &&`, `cache &`             | `void`                               |
+ *
+ * - init_origin_cell:
+ *
+ *     Implements the initialisation of the matrix origin \f$ (M(0,0)) \f$. This function gets the current cell
+ *     dereferenced from the score matrix and the alignment algorithm cache.
+ *
+ * - init_column_cell:
+ *
+ *     Implements the initialisation of the cells in the first row of the matrix \f$ (M(i,0)) \f$.
+ *     This function gets the current cell dereferenced from the score matrix and the alignment algorithm cache.
+ *
+ * - init_row_cell:
+ *
+ *     Implements the initialisation of the cells in the first column of the matrix \f$ (M(0,j)) \f$.
+ *     This function gets the current cell dereferenced from the score matrix and the alignment algorithm cache.
+ *
+ * ### Existing gap init policies:
+ *
+ *  - seqan3::detail::affine_gap_init_policy
```

Comment:
Grwat explanation! But is this really a costumisation point, i.e. can users just add their own policy, somehow squeeze it into the config and everything will work?

Maybe you could make it a little clearer what part of this documentation to refers to users adding custom behaviour in their app and which part only addresses developers of the library (most policies seem to be detail anyway?). And maybe mark respective things with `\if DEV` to it is only visible in the developer documentation?

Reply:
Oh, indeed it needs to be put into DEV. I forot this.

The intention was to provide an easier way to customise the alignment algorithms for library developers. Because most changes will involve structural changes in the algorithm and there
are depdencies between the data structures and layouts changing one thing means adapting others. Hence, the list of policies, what they are for, which interfaces they have to obey and that I need some more descriptive classes rather than plain tuples to interface with the different policies. So there are dpendencies but in general they are designed such that they can be replaced by some other implementation. That's why I would like to keep them noted as customisation points.